### ISSUE TITLE
[Integration Proposal] Proposal to add CodeSpecs library to core modules list

### ISSUE THREAD

**Issue by prowwid** [Codeception/Codeception#4407]
I'm developing [CodeSpecs](https://github.com/dekeysoft/code-specs) library that represents a Codeception module for advanced unit testing.
It is a fork of my initial library [PHPUnit Tester](https://github.com/dekeysoft/pu-tester) for PHPUnit with a lot of improvements and tuned to work with Codeception.

The idea of CodeSpecs is to think of a test case as of a behavior specification for a specific class. 
Each tests methods represented as a spec of a specific method of an object and describes some part of an object behavior you want to test.
You can read more and see for yourself how it looks like in the [documentation](https://github.com/dekeysoft/code-specs/tree/master/docs)

I've talked about the realization of this module with @DavertMik last year but didn't had time to implement anything then. Current realization is still very rough but I've already implemented general ideas of a library.

_What do you think about [CodeSpecs](https://github.com/dekeysoft/code-specs) module?_
_Is there a possibility to integrate it to the Codeception in the future versions (maybe next major release) or at least have it listed in docs?_

Looking forward to your comments.

**Comment by Naktibalda** [comment]
Hint: you should add your module to https://github.com/Codeception/codeception.github.com/edit/master/_data/modules.yml right away and it will appear in http://codeception.com/addons

Requirement of PHP 7.1 really too high for a core module.
We will catch up with you in 2-3 years.

**Comment by prowwid** [comment]
Hello @Naktibalda , thanks for the reply.

> Requirement of PHP 7.1 really too high for a core module.
This is why I've suggested next major release(eventually you'll need to move forward to PHP 7.1 or even newer version). 
Anyway, language version is what can be discussed without problems - it's easy to revert PHP 7 code to PHP 5.*, yet I won't be happy about this.

**Comment by prowwid** [comment]
Hello @DavertMik  and @Naktibalda,

So if  I would lower code requirements for CodeSpecs to PHP 5.4 could we talk business? :blush:

I like the ideas of Codeception, but, from my point of view, right now it really missing the support of Unit tests. 
For many teams Unit tests as important as Acceptance and Functional and I would really like to work on getting Unit tests more friendly and beautiful on Codeception.:blush:

So what do you think?
Could we negotiate a collaboration or for me better to went rogue on my own?:smiling_imp:

**Comment by DavertMik** [comment]
Thanks @prowwid 
I really like the initial implementation. And yes, I can agree that we don't have anything cool for unit tests.

> I like the ideas of Codeception, but, from my point of view, right now it really missing the support of Unit tests.

However, I don't see any killer feature in CodeSpecs right now. Something I could say to everyone, "hey, that's better than standard PHPUnit and PhpSpec!" My attempts to do this was `verify` and `specify` libraries but they are still not good enough for me )

Do you have an example with a real business case tested with CodeSpecs. Maybe just as draft. 
I'd like to start further discussion, maybe we can get some good ideas

**Comment by prowwid** [comment]
>However, I don't see any killer feature in CodeSpecs right now. 
Because I haven't added any yet:joy:
To make something really cool I would need deeper integration with Codeception core and this can be achieved by integration CodeSpecs to the core modules. 

I can't show you the exact code we use in tests because of NDA but I can sow some examples with edited messages and code to show the approximate look. With PHPUnit tester, we have in  tests code like:
```php
        $tester = $this->tester;
        $tester->checksScenario('clear relation when entity goes back to the "Entities Queue" tab');
        $tester->expectsThat('entity was saved successfully')->boolean($entity->saveUsingTransaction())->isTrue();
        $tester->expectsThat('entity relation was cleared')->valueOf($entity->ref_relation)->isNull();
    
```

```php
        $this->tester->checksScenario('generating validation for field')
            ->expectsThat('entity generates {censor} validation rule for {censor} based on {censor}')
            ->valueOf($entity->getRule())
            ->isEqualTo(static::RULE);
```
Unfortunately I can't show you more complex scenarios bacause it would reveal business rules and this is forbidden for me.

AS for CodeSpecs, I haven't released stable version (I don't know how the hell 0.0.1 get there - probably I accidentally added it) so there are no prroduction examples yet.

**Comment by prowwid** [comment]
Hi @DavertMik,
thanks for a quck reply. 

>However, I don't see any killer feature in CodeSpecs right now. 

Because I haven't added any yet:joy:
To make something really cool I would need deeper integration with Codeception core and this can be achieved by integration CodeSpecs to the core modules. 
Even to have test names like `UserSpec` or `UserBehavior` I would need to add an additional loader and edit `Codeception\Test\Loader` in the core to use the additional loader for new tests format.
This is I so eagerly want to have CodeSpecs in the core as this is the only way to have the best integration with the framework and be able to do cool stuff:blush:
CodeSpecs right now can do what PHPUnit Assert can but I don't want to stop on this - I would extend the functionality but need to understand whether CodeSpecs would be an independent module or integrated to the Codeception first.

> Something I could say to everyone, "hey, that's better than standard PHPUnit and PhpSpec!" 

As for PhpSpec, it is a completely different framework. Don't you agree that having a single framework for all kinds of tests is better than having one that does everything cool?:smile:
Also, one of the reasons we do not like PhpSpec is method naming. 
In RSpec `it "should save user if he is active and rejects otherwise"` looks cool, but, from my point of view,  PhpSpec `function it_should_save_user_if_he_is_active_and_rejects_otherwise` doesn't look cool at all.
But, again, it is my opinion and better leave PhpSpec alone for now:blush:

Please don't get me wrong, but I think CodeSpecs already better than PHPUnit Asserts:smile:
As I see it, for example:
```php
$I->seeTahtBoolean('result of entity validation', $result)->isTrue();
```
looks better and much friendly than
```php
$this->assertTrue($result);
```

The reason I've developed "PHPUnit Tester" in the first place were asserts like:
```php
$this->assertTrue($result);
$this->assertEquals($result, 5);
$this->assertArrayHasKey('value', $result);
```  
When test fails and I see error message with something like "Failed asserting that false is true" and then go to test and see `$this->assertTrue($result);` it is a real pain, especially if `$result` is returned by some class with complex  logic and even to understand why there should have been `true` in the first pace you need to learn a lot of things.
So my primary goal still is to achieve readable tests with friendly output to console if something goes wrong. 
I knew people that don't like tests only because don't see how they would help them if they would see something like "Failed asserting that false is true".

>Do you have an example with a real business case tested with CodeSpecs. 

I haven't released stable version (I don't know how the hell 0.0.1 get there - probably I accidentally added it) so there are no production examples yet.

Right now I have only PHPUnit Tester examples.

But I can't show you the exact code we use in tests because of NDA but I can sow some examples with edited messages and code to show the approximate look. With PHPUnit tester, we have in  tests code like:
```php
        $tester = $this->tester;
        $tester->checksScenario('clear relation when entity goes back to the "Entities Queue" tab');
        $tester->expectsThat('entity was saved successfully')->boolean($entity->saveUsingTransaction())->isTrue();
        $tester->expectsThat('entity relation was cleared')->valueOf($entity->ref_relation)->isNull();
    
```

```php
        $this->tester->checksScenario('generating validation for field')
            ->expectsThat('entity generates {censor} validation rule for {censor} based on {censor}')
            ->valueOf($entity->getRule())
            ->isEqualTo(static::RULE);
```
Unfortunately, I can't show you more complex scenarios because it would reveal business rules and this is forbidden for me.

Even with PHPUnit Tester, we have a more readable test right now and if something fails we can identify the cause a lot quicker and easier. But with CodeSpecs testing would be a lot more fun and friendly:wink:

**And here is a rough example of how a business case might look like in CodeSpecs:**
````
class IncomeCalculatorBehavior extends BehaviorSpecification {
    private const EXPECTED_TAX_FOR_FIRST_LEVEL_TAX_RULE = 4500;
    private const EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE = 7200;
    private const EXPECTED_TAX_FOR_THIRD_LEVEL_TAX_RULE = 30000;

    /**
     * @test
     */
    public function calculateTaxSpec() {
        $service = new IncomeService($clientsPayments, $hoursSpentWorking);
        $I = $this->tester;
        $I->describe('income tax calculations');

        $I->verifyThat('income calculator honors tax rules for different ranges of income');

        $I->expectThat('for income less that 50 000 calculator use 10% tax rule');
        $I->seeThatNumber('income tax', $service->calculateTax())
            ->isNotEmpty()
            ->isEqualTo(self::EXPECTED_TAX_FOR_FIRST_LEVEL_TAX_RULE);

         // create new calculator here
        $I->expectThat('for income between 50 000 and 100 000 calculator use 12% tax rule');
        $I->seeThatNumber('income tax', $service->calculateTax())
            ->isNotEmpty()
            ->isEqualTo(self::EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE);

        // create new calculator here
        $I->expectThat('for income more than 100 000 calculator use 20% tax rule');
        $I->seeThatNumber('income tax', $service->calculateTax())
            ->isNotEmpty()
            ->isEqualTo(self::EXPECTED_TAX_FOR_THIRD_LEVEL_TAX_RULE);
    }
}
````
**Note:** class with `IncomeCalculatorBehavior` won't be executed without additional loader, I just want to show you how it might look like. 

@DavertMik it is a rough example made from my mind so if you have a better scenario in mind I can make a better example.

With best regards,
Dmitry

**Comment by prowwid** [comment]
BTW, here is an example of output for failed spec from `IncomeCalculatorBehavior `:
![failed spec](https://user-images.githubusercontent.com/3070281/28464336-5c622f2e-6e2d-11e7-963e-532aad574fb6.png)

**Comment by Naktibalda** [comment]
> Even to have test names like UserSpec or UserBehavior I would need to add an additional loader and edit Codeception\Test\Loader in the core to use the additional loader for new tests format.

I've seen this coming up before. @DavertMik How about opening up the core to allow loading of custom test formats?

**Comment by Naktibalda** [comment]
```php
$I->seeThatNumber('income tax', $service->calculateTax())
             ->isNotEmpty()
             ->isEqualTo(self::EXPECTED_TAX_FOR_THIRD_LEVEL_TAX_RULE);
```

@prowwid Do you really need a custom test format?
I don't see anything special in your example.
`seeThatType` methods could be implemented as a module.

**Comment by prowwid** [comment]
Hi @Naktibalda,
 
>Do you really need a custom test format?
I don't see anything special in your example.
seeThatType methods could be implemented as a module.

Sorry, didn't get what you meant. Could you please explain more detailed?

**Comment by Naktibalda** [comment]
Have you tried to implement your methods as a module?

**Comment by prowwid** [comment]
@Naktibalda ,

>Have you tried to implement your methods as a module?

What would be the benefit of having methods as modules? Won't it be an overhead of modules?

**Comment by Naktibalda** [comment]
I mean one module for CodeSpec, not one module for each method.
No modifications of Codeception core are necessary for implementation of the module.

**Comment by prowwid** [comment]
@Naktibalda ,

CodeSpecs right now is a prototype but is already a Codeception module.
I'm not really following what else could/should I implement as a  module.

**Comment by Naktibalda** [comment]
Why do you need a custom file format?

**Comment by prowwid** [comment]
@Naktibalda because it changes the conceptual meaning of a test.

One of the primary goals I want to achieve is to make conceptual meaning of a test more friendly. 
All of this methods `describe`, `verifyThat`, `expectThat`, `seeThat*` designed to change a meaning of a test from some purely techical "cehck this, check taht" to a user-friendly format that being written in a human readable way.
The idea of dropping "test" in order to use "spec" keyword is to change the meaning of a test so that developer would see it as a specification of a specific class and even use it as some kind of a documentation.

Basically something like `$I->seeTahtBoolean('result of entity validation', $result)->isTrue();` is the same as `$this->assertTrue($result);` from technical point of view. 
But from conceptual point of view, it is completely different things and UX is better in the first example(hope you get what I'm trying to say:smile:) 

People love Codeception because it changed the conceptual meaning of functional and acceptance tests and allowed us to use a bunch of cool, friendly methods to express what we want to test. I want to achieve the same for unit tests.

**Comment by prowwid** [comment]
Hi @DavertMik ,

Do you have any new thoughts on this one?

Looking forward to your comments,

With best regards,
Dmitry

**Comment by DavertMik** [comment]
@prowwid sorry, I was working on CodeceptJS all the last week. Need to switch my mind back to PHP to reply )

**Comment by DavertMik** [comment]
Ok, from technical side it looks like you the best is to use Cest format for unit testing in this manner. It was mentioned in docs but I removed it (stupid me). The current code can already work in Cest format if you pass `$I` into it. I don't like the name `Cest` either, maybe it should be renamed to Spec one day as well. But maybe we need another word which is not so associated with tools like PhpSpec... Not sure

Ok. I agree `assertTrue(false)` is meaningless, descriptions are cool and they should be used widely. However, we should not forget that the main point of the unit test is the business code. It should be as close to its real usage as possible. In case of your example, there are just too much descriptions. They make test a bit more complicated than it is: mixing of code and comments is happening which will affect the writing of complex tests. I have few proposals how to improve current tests:

1. move global descriptions to annotations (they are just comments, right?)
2. make the actual code more visible, don't hide it in depth of `$I->seeThat`... call

```php
/**
 * @describe income tax calculations
 * @verify income calculator honors tax rules for different ranges of income
 */
public function calculateTaxSpec(UnitTester $I) {

    $service = new IncomeService($clientsPayments, $hoursSpentWorking);

    $I->expectThat('for income less that 50 000 calculator use 10% tax rule');
    $I->seeThat($service->calculateTax())
        ->is('incoming tax')
        ->isNumber()
        ->isEqualTo(15);

        // create new calculator here
    $I->expectThat('for income between 50 000 and 100 000 calculator use 12% tax rule');
    $I->seeThatNumber('income tax', $service->calculateTax())
        ->isNotEmpty()
        ->isEqualTo(self::EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE);

    // create new calculator here
    $I->expectThat('for income more than 100 000 calculator use 20% tax rule');
    $I->seeThatNumber('income tax', $service->calculateTax())
        ->isNotEmpty()
}
```

We can even simplify the code to add realtime custom matcher (to reduce the code):

```php
/**
 * @describe income tax calculations
 * @verify income calculator honors tax rules for different ranges of income
 */
public function calculateTaxSpec(UnitTester $I) {
    $assertTax = $I->match('incoming tax')->isNumber()->isNotEmpty();

    $service = new IncomeService($clientsPayments, $hoursSpentWorking);

    $I->expect('for income less that 50 000 calculator use 10% tax rule');
    $assertTax($service->calculateTax())->isEqualTo(15);

    $I->expectThat('for income between 50 000 and 100 000 calculator use 12% tax rule');
    // create new calculator here
    $assertTax($service->calculateTax())->isEqualTo(self::EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE);

    $I->expectThat('for income more than 100 000 calculator use 20% tax rule');
    // create new calculator here
    $assertTax($service->calculateTax())->isEqualTo(17);
}
```

I think that those custom matchers could probably be the cool feature. There is no such thing in PHPUnit

**Comment by Naktibalda** [comment]
> I think that those custom matchers could probably be the cool feature. There is no such thing in PHPUnit

Have you seen [Hamcrest for PHP](https://github.com/hamcrest/hamcrest-php)?

**Comment by prowwid** [comment]
Hi @DavertMik ,

Thanks for the reply.

> At least you don't need that ugly `$this->tester` there. 

I absolutely agree with you. I don't like using `$this->tester` either. And would prefer the following format:
```php
public function calculateTaxSpec(UnitTester $I) {
```

Right now I use `$this->tester`  only for compatibility with PHPUnit. For example, right now I don't know how to pass tester as a first argument without patches to core and not affect PHPUnit's [Data Providers](https://phpunit.de/manual/current/en/writing-tests-for-phpunit.html#writing-tests-for-phpunit.data-providers) functionality.
**_Maybe you know how to do this or I'm right that it's impossible without a patch to the core?_**

>move global descriptions to annotations (they are just comments, right?)

Yeah, `@describe` could be moved to annotations. But I don't think it's a place for 'verify'. 
You see, my idea of "verify" is to group several "expectThat" into a meaningful group. So you can have several "verifyThat" in a spec.
The structure with "verifyThat" might look like:
```
describe - high level
    verifyThat - optional
        expectThat
        expectThat
        expectThat
    verifyThat
        expectThat
        expectThat
...........
```

The structure without "verifyThat" might look like:
```
describe - high level
    expectThat
    expectThat
    expectThat
...........
```

I think it is redundantly having two annotations with messages.


>We can even simplify the code to add realtime custom matcher (to reduce the code):

I like this idea. Would work on implementing this approach.

>make the actual code more visible

I see what you mean. But in your example 

```php
$I->seeThat($service->calculateTax())
        ->is('incoming tax')
```

I don't like `->is(`.  "is" makes sense in matchers  like "isNumber", "isEmpty", "isEqualTo" because "is" represents a state of a value. But I don't think it's suited to describe what is being matched.
 
I think, maybe something like:
 ```php
$I->lookAt('incoming tax')
        ->seeThat($service->calculateTax())
        ->isEqualTo(self::EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE);
``` 
would be better?

And if you don't want to have a custom name of a variable you can just use:
 ```php
$I->seeThat($service->calculateTax())
        ->isEqualTo(self::EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE);
``` 

What is your opinion?

**Comment by DavertMik** [comment]
>  For example, right now I don't know how to pass tester as a first argument without patches to core and not affect PHPUnit's Data Providers functionality.

Cest format supports data providers. 

> verifyThat

verifyThat looks a lot like [specify](https://github.com/codeception/specify) then. So you can wrap the code into closure and make it visible. You can even continue execution if one of verifyThat blocks failed:

```php
$I->verifyThat('xxxx', function($I) {
   .....
});
```

> lookAt

this name is still not good enough... other ideas?

**Comment by prowwid** [comment]
Hi @DavertMik ,

> Cest format supports data providers.

Sorry, I probably misunderstood your initial message about Cests. So you are suggesting to get rid of base unit class and have specs looks like cests? For example:
```php
class IncomeCalculatorSpec {

    public function calculateTaxSpec(Tester $I) {
        ......    
    }
}
```
This way would be better, I agree. 

>verifyThat looks a lot like specify then.

I don't really like "specify" format. First of all because when it comes to complex testing you age complicated callbacks and really long lines of code that makes the test harder to maintain and read. Still, it was my cases that might not be common(I'm working with large applications that being developed by years).
And the second reason is that "specify" is messing with tests behavior and in IDE I often get incorrect reports - IDE shows that tests still running (yet, I believe, this issue can be handled and might have already had but I don't use specify anymore so can't tell).
Anyway, it's only my personal opinion.  If you think that "specify" format is better I won't argue - in many cases, it looks really cool:blush:

>this name is still not good enough... other ideas?

Yeah, I understand. I should have phrased it differently. I wanted to hear your opinion about using a method, that describes what are you going to check, that goes before "seeThat" but not after.

What do you think about following names:
```php
$I->amGoingToCheck('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->amGoingToTest('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->amGoingToReview('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->wantToCheck('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->wantToTest('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->wantToReview('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->amChecking('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->amReviewing('incoming tax')
   ->seeThat($service->calculateTax())
```
```php
$I->amTesting('incoming tax')
   ->seeThat($service->calculateTax())
```

The better method name is what I can research. But what do you think about the approach? 
Would it be good enough having an optional method that can be called before "seeThat" to describe the value that would be tested?
Or I need to come up with a better approach?

**Comment by prowwid** [comment]
Hello @DavertMik ,

I've just made a few changes in CodeSpecs that include:

- new method `$I->match('variable name')` to create a real-time custom matcher
- new method `$I->amGoingToCheck('variable name)` to specify a name of a variable that would be checked by matcher
- new method `$I->seeThat($variable)` to create a matcher

Example with all bored IncomeCalculator that shows usage of a new methods:
```php
        $I->describe('income tax calculations');

        $matchIncomeTax = $I->match('income tax')->isNumber()->isNotEmpty()->isGreaterThan(0);

        $I->verifyThat('income calculator honors tax rules for different ranges of income');

        $I->expectThat('for income less that 50 000 calculator use 10% tax rule');
        $matchIncomeTax($service->calculateTax())
            ->isEqualTo(self::EXPECTED_TAX_FOR_FIRST_LEVEL_TAX_RULE);

        $I->expectThat('for income between 50 000 and 100 000 calculator use 12% tax rule');
        $matchIncomeTax( $service->calculateTax())
            ->isEqualTo(self::EXPECTED_TAX_FOR_SECOND_LEVEL_TAX_RULE, 'expected tax of a second level tax');

        $I->expectThat('for income more than 100 000 calculator use 20% tax rule');
        $I->amGoingToCheck('income tax');
        $I->seeThat( $service->calculateTax())
            ->isNumber()
            ->isNotEmpty()
            ->isEqualTo(self::EXPECTED_TAX_FOR_THIRD_LEVEL_TAX_RULE, 'expected tax of a third level tax');
    
```

Here is the [link to example in the repository](https://github.com/dekeysoft/code-specs/blob/master/tests/Unit/Example/IncomeCalculatorTest.php)

Looking forward to your comments.

With best regards,
Dmitry

**Comment by prowwid** [comment]
Hello @DavertMik ,

do you have any new thoughts on this one?
I would really like your opinion on last comments to choose the next step to proceed with

With best regards,
Dmitry

**Comment by prowwid** [comment]
Hi everyone,

It's more than a month since the last comment in this task. 
Could anyone please say at least something? I really need your opinion to proceed with implementations.

With best regards,
Dmitry

**Comment by Naktibalda** [comment]
My opinion is that there is no need to include it to Codeception.

**Comment by DavertMik** [comment]
Yeah, sorry for the delay, I was trying to rethink it few times.

This doesn't correspond to my current style of testing. So I won't be able to promote and persuade it. 
For unit/integration tests I try to write as less redundant code as possible to keep maintainability, to keep the closeness to codebase.

I think this should be part of a standalone module. I could mention it in docs for Unit Testing but I agree we are not going to sign to maintain this module as a part of Codeception.
